sdk/src/utils/pdf_utils.rs

```diff
@@ -0,0 +1,403 @@
+
+pub(crate) trait C2paPdf: Sized {
+    /// Load a PDF from a slice of bytes.
+    fn from_bytes(bytes: &[u8]) -> Result<Self, Error>;
+
+    /// Save the `C2paPdf` implementation to the provided `writer`.
+    fn save_to<W: Write>(&mut self, writer: &mut W) -> Result<(), std::io::Error>;
+
+    /// Returns `true` if the `PDF` is password protected, `false` otherwise.
+    fn is_password_protected(&self) -> bool;
+
+    /// Returns `true` if this PDF has `c2pa` Manifests, `false` otherwise.
+    fn has_c2pa_manifest(&self) -> bool;
+
+    /// Writes provided `bytes` as a PDF `Embedded File`
+    fn write_manifest_as_embedded_file(&mut self, bytes: Vec<u8>) -> Result<(), Error>;
+
+    /// Writes provided `bytes` as a PDF `Annotation`.
+    fn write_manifest_as_annotation(&mut self, vec: Vec<u8>) -> Result<(), Error>;
+}
+
+pub(crate) struct Pdf {
+    document: Document,
+}
+
+impl C2paPdf for Pdf {
+    fn from_bytes(bytes: &[u8]) -> Result<Self, Error> {
+        let document = Document::load_mem(bytes)?;
+        Ok(Self { document })
+    }
+
+    /// Saves the in-memory PDF to the provided `writer`.
+    fn save_to<W: Write>(&mut self, writer: &mut W) -> Result<(), std::io::Error> {
+        self.document.save_to(writer)
+    }
+
+    fn is_password_protected(&self) -> bool {
+        self.document.is_encrypted()
+    }
+
+    /// Determines if this PDF has a C2PA manifest embedded.
+    ///
+    /// This is done by checking if the Associated File key of the catalog points to a `Dictionary`
+    /// with an `AFRelationship` set to `C2PA_Manifest`.
+    fn has_c2pa_manifest(&self) -> bool {
+        self.document
+            .catalog()
+            .and_then(|catalog| catalog.get_deref(ASSOCIATED_FILE_KEY, &self.document))
+            .and_then(Object::as_dict)
+            .and_then(|dict| dict.get_deref(AF_RELATIONSHIP_KEY, &self.document))
+            .and_then(Object::as_name)
+            .map(|name| name == C2PA_RELATIONSHIP)
+            .unwrap_or_default()
+    }
+
+    /// Writes the provided `bytes` to the PDF as an `EmbeddedFile`.
+    fn write_manifest_as_embedded_file(&mut self, bytes: Vec<u8>) -> Result<(), Error> {
+        // Add `FileStream` and `FileSpec` to the PDF.
+        let file_stream_ref = self.add_c2pa_embedded_file_stream(bytes);
+        let file_spec_ref = self.add_embedded_file_specification(file_stream_ref);
+
+        self.set_af_relationship(file_spec_ref)?;
+
+        let mut manifest_name_file_pair = vec![
+            Object::string_literal(CONTENT_CREDS),
+            Reference(file_spec_ref),
+        ];
+
+        if let Ok(names) = self.document.catalog_mut()?.get_mut(NAMES_KEY) {
+            // Determine if `/Names` value is direct or indirect.
+            let names = if let Ok(names) = names.as_reference() {
+                self.document.get_object_mut(names)?.as_dict_mut()?
+            } else {
+                names.as_dict_mut()?
+            };
```

Comment:
This pattern occurs a few times throughout the PR but has proved hard to abstract. `lopdf` doesn't have a great way to follow references when interacting with a mutable `lopdf` document. This is because when following references, you need a mutable reference to the document object, and a mutable reference to the indirect object you are attempting to get, which means you have two mutable references to `Document`.  I'm working on switching this code to make use of `ObjectIds` instead of the mutable reference. This is proving difficult for other reasons!

Reply:
Additionally, an issue with this pattern is that there's no guarantee there will only be one reference between objects. `lopdf` has implemented dereferencing helpers when using immutable references, but doesn't have anything for mutable references.

Reply:
Okay, will merge!